dali/test/python/test_pipeline.py

```diff
@@ -1134,6 +1134,117 @@ def iter_setup(self):
     assert(iter_num == i)
 
+def test_external_source_fail():
+    class ExternalSourcePipeline(Pipeline):
+        def __init__(self, batch_size, external_s_size, num_threads, device_id):
+            super(ExternalSourcePipeline, self).__init__(batch_size, num_threads, device_id)
+            self.input = ops.ExternalSource()
+            self.batch_size_ = batch_size
+            self.external_s_size_ = external_s_size
+
+        def define_graph(self):
+            self.batch = self.input()
+            return [self.batch]
+
+        def iter_setup(self):
+            batch = np.zeros([self.external_s_size_,4,5])
+            self.feed_input(self.batch, batch)
+
+    batch_size = 3
+    pipe = ExternalSourcePipeline(batch_size, batch_size - 1, 3, 0)
+    pipe.build()
+    try:
+        pipe.run()
+    except RuntimeError:
+        pass
+    else:
+        assert False, "ExternalSource should throw"
+
+def test_external_source_fail_list():
+    class ExternalSourcePipeline(Pipeline):
+        def __init__(self, batch_size, external_s_size, num_threads, device_id):
+            super(ExternalSourcePipeline, self).__init__(batch_size, num_threads, device_id)
+            self.input = ops.ExternalSource()
+            self.batch_size_ = batch_size
+            self.external_s_size_ = external_s_size
+
+        def define_graph(self):
+            self.batch = self.input()
+            return [self.batch]
+
+        def iter_setup(self):
+            batch = []
+            for _ in range(self.external_s_size_):
+                batch.append(np.zeros([3,4,5]))
+            self.feed_input(self.batch, batch)
+
+    batch_size = 3
+    pipe = ExternalSourcePipeline(batch_size, batch_size - 1, 3, 0)
+    pipe.build()
+    try:
+        pipe.run()
+    except RuntimeError:
+        pass
+    else:
+        assert False, "ExternalSource should throw"
+
+def test_external_source_scalar():
+    class ExternalSourcePipeline(Pipeline):
+        def __init__(self, batch_size, external_s_size, num_threads, device_id, label_data):
+            super(ExternalSourcePipeline, self).__init__(batch_size, num_threads, device_id)
+            self.input = ops.ExternalSource()
+            self.batch_size_ = batch_size
+            self.external_s_size_ = external_s_size
+            self.label_data_ = label_data
+
+        def define_graph(self):
+            self.batch = self.input()
+            return [self.batch]
+
+        def iter_setup(self):
+            batch = []
+            for i in range(self.external_s_size_):
+                batch.append(np.array(self.label_data_ + i, dtype=np.uint8))
+            self.feed_input(self.batch, batch)
+
+    batch_size = 3
+    label_data = 10
+    pipe = ExternalSourcePipeline(batch_size, batch_size, 3, 0, label_data)
+    pipe.build()
+    for _ in range(10):
+      out = pipe.run()
+      for i in range(batch_size):
+          assert out[0].as_array()[i] == label_data + i
+
+def test_external_source_list():
```

Comment:
This tests seems to be exactly the same as the one above (`test_external_source_scalar`) except the print in line 1236

Reply:
It was suppose to be list in one place and the scalar in the second. I just reworked it to be more verbose.